glean-core/src/database/mod.rs

```diff
@@ -495,15 +491,20 @@ impl Database {
             return Ok(());
         }
 
-        // Lifetime::Ping data is not persisted to disk if
-        // Glean has `delay_ping_lifetime_io` set to true
+        // Lifetime::Ping might also have data saved to `ping_lifetime_data`
+        // in case `delay_ping_lifetime_io` is set to true
         if lifetime == Lifetime::Ping {
             if let Some(ping_lifetime_data) = &self.ping_lifetime_data {
                 let mut data = ping_lifetime_data
                     .write()
                     .expect("Can't access app lifetime data as writable");
-                data.remove(&final_key);
-                return Ok(());
+                if data.remove(&final_key).is_some() {
+                    // If the metric is found in `ping_lifetime_data` there is
+                    // no need to check if it is also in rkv. Data should always
+                    // be either in one place or the other, since we clear
+                    // `ping_lifetime_data` upon persisting it to rkv.
```

Comment:
I think this might lead to invalid data.

Consider this:

1. Ping-lifetime histogram data gets recorded, its stored in memory.
2. Ping-lifetime data is persisted, memory is cleared.
3. More ping-lifetime histogram data gets recorded, its stored in memory.
	* **NOTE**: This now starts from a fresh histogram
4. Ping-lifetime data is collected and send in a ping
    * **NOTE**: We're losing the first round of writes.

Reply:
Oh, great catch there.

Reply:
Alright, right now how this works is: data is always **either** in memory or persisted. All the read / delete / update functions will check in both places and the create functions will always create in memory. Persistence will happen on demand and clear memory. I now understand how that can create problems.

The solution to these problems would be to change the way that works to something like: in memory data will mirror persisted data + any new changes and create / read / delete / update function will only apply to in memory data. Persisted data will then be updated on demand, but in memory will not be cleared.

> Is loading persisted data on startup coming in a follow up?

The way it is right now, I don't think this applies. With the new solution, yes that makes sense.

This is going to be a major change to the code on this PR. I think it might make sense to close this one and create a new one when I'm done. What  do you think?

Reply:
I'm fine with re-using this PR, just force push if needed.
Though if it is easier for your own workflow starting fresh could work just as well, I don't mind either way.

The loading-on-startup part can be done in a follow up or with the other changes, I also don't mind that much.